File: 86-partition-list/partition-list.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function partition(head: ListNode | null, x: number): ListNode | null {
    if (!head) return null;

    const leftNodes: ListNode[] = [];
    const rightNodes: ListNode[] = [];

    let curr = head;
    while (curr) {
        if (curr.val < x) {
            leftNodes.push(curr);    
        } else {
            rightNodes.push(curr);
        }
        const nextNode = curr.next;
        curr.next = null;
        curr = nextNode;
    }

    
    let leftHead: ListNode = new ListNode(null); 
    let leftTail: ListNode = leftHead
    for (const node of leftNodes) {
        leftTail.next = node;
        leftTail = node;
    }

    let rightHead: ListNode = new ListNode(null);
    let rightTail: ListNode = rightHead;
    for (const node of rightNodes) {
        rightTail.next = node;
        rightTail = node;
    }

    if (!leftHead.next) {
        leftHead.next = rightHead.next
    } else {
        leftTail.next = rightHead.next;
    }

    return leftHead.next
};

File: 86-partition-list/partition-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        if not head: return None

        leftNodes = []
        rightNodes = []

        curr = head
        while curr:
            if curr.val < x:
                leftNodes.append(curr)
            else:
                rightNodes.append(curr)
            nextNode = curr.next
            curr.next = None
            curr = nextNode
        
        leftHead = ListNode(None)
        leftTail = leftHead
        for node in leftNodes:
            leftTail.next = node
            leftTail = node
        
        rightHead = ListNode(None)
        rightTail = rightHead
        for node in rightNodes:
            rightTail.next = node
            rightTail = node
        
        if not leftHead.next:
            leftHead.next = rightHead.next
        else:
            leftTail.next = rightHead.next
        
        return leftHead.next